R Syntax: Load QML Controls as QML Module

boutinb opened the pull request:
Make the QML Components as a QML Module (by using the qt_add_qml_module cmake function).
This makes also possible to use this module as a plugin: this is needed for the R Syntax project

All versions in QML import statement must be removed.

No qmldir needed (it is generated automatically), and no need to register all the base QML objects: they are automatically registered via the QML_ELEMENT macro.

~~The module is linked to the CommonData and has access to all DataSet objects, but it should not use static members (as the sqlite database is instantiated in another translation unit, any call to this instance would crash). So a DatasetInterface must be created to load the sqlite database, and a new DataSet must use this DatasetInterface object: (as in the Engine). The Audit analyses creates Filter object: this object must be created via this DataSet object.
The Consumer/Provider pattern that was used to read data for the QML controls, could be in the future replaced by calls to this DataSet object.~~

boutinb committed:
Load QML Controls as QML Module

The QML Module is also a plugin.

All versions in QML import statement must be removed.

boutinb committed:
More

shun2wang commented on QMLComponents/plugin.cpp:
```diff
@@ -0,0 +1,62 @@
+
+		QLocale::setDefault(QLocale(QLocale::English)); // make decimal points == .
+
+		QmlUtils::setGlobalPropertiesInQMLContext(engine->rootContext());
```
Build on Linux failed, see: https://github.com/jasp-stats/jasp-desktop/actions/runs/12710965693/job/35433254997?pr=5776#step:9:230

boutinb replied:
Oh I see, thanks for the warning!

boutinb committed:
Remove Filter object from QML Components

Remove the CommonData to the QML Comnponent.
Only the Audit analyses use the CommonData objects (via the Filter object), but this leads to unexpected error: the sqlite library is added in the translation unit of the QML Component library.
Furthermore, n order to use the Filter object in the QML Component the whole DataSet object was exposed to the QML Component, which is completely inconsistent with the Provider/Consumer pattern used in the VariableInfo.
So in order to still be able to use the Filter object in the Audit analyses, callback functions are used instead, which are set in the DesktopCommunicator object.

boutinb committed:
Update submodule so that it uses QML import without version

shun2wang commented:
Sorry, but version things still not be removed in `Description.qml` `Upgrades.qml` in submodules I think?

boutinb committed:
JASP QML Controls base objects should be registered automatically

Use QML_ELEMENT
Remove import version in other QML files

boutinb commented:
> Sorry, but version things still not be removed in `Description.qml` `Upgrades.qml` in submodules I think?

That's true, I forgot these ones also.

JorisGoosen commented on CommonData/filter.cpp:
```diff
@@ -3,6 +3,8 @@
 #include "dataset.h"
 #include "databaseinterface.h"
 
+std::map<std::string, Filter*>	_filterMap;
```
This is a terrible Idea, a global really?

boutinb replied:
OK, I have changed this, by creating a new DatabaseInterface that loads the sqlite database. This is indeed much cleaner.

boutinb committed:
Add ALTNavigation

boutinb committed:
Use CommanData in QML Components

boutinb committed:
Final touch

JorisGoosen commented on CommonData/dataset.cpp:
```diff
@@ -619,3 +619,9 @@ bool DataSet::initColumnWithStrings(int colIndex, const std::string & newName, c
 
+bool DataSet::isFilterNameFree(const std::string& filterName)
+{
+	return -1 == db().filterGetId(filterName);
```
Makes more sense to have this on `Filter` because the name is independent of the DataSet in question.

JorisGoosen replied:
Why did you make it non-static?

JorisGoosen commented on CommonData/filter.cpp:
```diff
@@ -170,11 +170,6 @@ bool Filter::checkForUpdates()
 
-bool Filter::filterNameIsFree(const std::string &filterName)
-{
-	return -1 == DatabaseInterface::singleton()->filterGetId(filterName);
```
Which is also where it was...

JorisGoosen commented on QMLComponents/variableinfo.cpp:
```diff
@@ -68,5 +70,12 @@ bool VariableInfo::dataAvailable()
 DataSet *VariableInfo::dataSet()
 {
-	return _provider ? reinterpret_cast<DataSet*>(_provider->provideInfo(VariableInfo::DataSetPointer).value<void*>()) : nullptr;
+	static DataSet* singleDataSet = nullptr;
```
This is not going to work is it?
How do you reflect the dataset getting destroyed?

JorisGoosen replied:
Yeah, so as I suspected:
- open dataset, use a sampling audit analysis, it works
- close dataset, dont close jasp
- open other dataset, use a sampling audit analysis, its broken...

Ill just make sure the datasetpointer works, or Ill fix it some other way

boutinb replied:
Hmm I forgot this case.

JorisGoosen committed:
use DataSetPointer again, jusat make sure DatabaseInterface is loaded

JorisGoosen committed:
_singleton should be accessed from cpp to make sure it is in the right context

JorisGoosen committed:
undo ColumnUtils::doubleToString* -> Utils::

JorisGoosen committed:
undo some more changes and move init of databaseinterface singleton to inside singleton() (but without creating the actual tablestructure, this remains the prerogative of DataSetPackage)

JorisGoosen committed:
small tweaks

JorisGoosen commented:
So ive reverted a bunch of the changes you made to cpp code that was unnecessary.
It now runs fine on macos, I will proceed to test Windows and Linux.

Commit:
try to get it to build on arch linux

linux compiles and runs
